### PR TITLE
Added support for custom themes

### LINKED ISSUE #3809: Feature Request: Custom Themes
**Is your feature request related to a problem? Please describe.**
When using custom themes for windows, or custom themes in general, you are unable to manually change the color values of the Files app.

**Describe the solution you'd like**
I would like a hidden/experimental settings panel from which you could change the hex values of the app.

**Describe alternatives you've considered**
Another solution would be a config file from which you could change the hex values of the app.

**Additional context**
Edit: I know this is similar to #3176, but I believe this to not be a duplicate because in #3176, the author wants support for custom explorer themes, whereas I want a way to manually change the color codes via a gui or a config file.

### PR DESCRIPTION
**Pull request by winston-de**
Closes #3809 
This PR adds base support for theming, and selecting themes.
Not to be confused with light/dark themes, as these themes can utilize both.

You can test this by adding themes to the themes folder that should be created when you run the branch.

Themes are in the form of xaml files with resource dictionaries.

Basic example of theming:
image

image

### CODE REVIEW

File: Files/Views/SettingsPages/Appearance.xaml [hunk 1]
```diff
@@ -43,6 +43,16 @@
...
 
+                <ComboBox
+                    x:Name="ColorSchemeChooser"
+                    Width="Auto"
+                    MinWidth="200"
+                    MaxWidth="250"
+                    Header="Color scheme"
+                    HeaderTemplate="{StaticResource CustomHeaderStyle}"
+                    ItemsSource="{Binding ColorSchemes}"
+                    SelectedItem="{Binding SelectedThemeName, Mode=TwoWay}" />
```
Comment (yaichenbaum) [inline Files/Views/SettingsPages/Appearance.xaml:54]: Is it posible to only show this combo box when Files detects custom themes in the directory?

Comment (winston-de) [reply Files/Views/SettingsPages/Appearance.xaml:54]: Ye

File: Files/Views/SettingsPages/Appearance.xaml [hunk 2]
```diff
@@ -43,6 +43,18 @@
...
+                    x:Name="ColorSchemeChooser"
+                    x:Uid="SettingsApperanceColorScheme"
+                    Width="Auto"
+                    MinWidth="200"
+                    MaxWidth="250"
+                    Header="Color scheme"
+                    HeaderTemplate="{StaticResource CustomHeaderStyle}"
+                    ItemsSource="{Binding ColorSchemes}"
+                    SelectedItem="{Binding SelectedThemeName, Mode=TwoWay}" 
+                    Visibility="{Binding ShowColorSchemeSelector}"/>
```
Comment (yaichenbaum) [inline Files/Views/SettingsPages/Appearance.xaml:56]:
```suggestion
                    x:Load="{Binding ShowColorSchemeSelector}"/>
```

**Comment by winston-de** [comment]
@yair100 ```x:load``` doesn't seem to support ```Binding```, and ```x:Bind``` doesn't seem to work with the datacontext. To get around this, I moved ```ShowColorSchemeSelector``` to code-behind. Is this ok?

**Comment by yair100** [comment]
> @yair100 `x:load` doesn't seem to support `Binding`, and `x:Bind` doesn't seem to work with the datacontext. To get around this, I moved `ShowColorSchemeSelector` to code-behind. Is this ok?

@winston-de That should be fine in this case.

File: Files/Helpers/ExternalResourcesHelper.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        };
+        
+        private StorageFolder ThemeFolder { get; set; }
+
+        public string CurrentThemeResources { get; set; }
+
+        public async Task LoadSelectedTheme()
+        {
+            ThemeFolder = (await ApplicationData.Current.LocalFolder.TryGetItemAsync("themes")) as StorageFolder;
+            ThemeFolder ??= await ApplicationData.Current.LocalFolder.CreateFolderAsync("themes");
```
Comment (yaichenbaum) [inline Files/Helpers/ExternalResourcesHelper.cs:30]:
```suggestion
            ThemeFolder = (await ApplicationData.Current.LocalFolder.TryGetItemAsync("Themes")) as StorageFolder;
            ThemeFolder ??= await ApplicationData.Current.LocalFolder.CreateFolderAsync("Themes");
```

File: Files/Strings/en-US/Resources.resw [hunk 4]
```diff
@@ -1908,4 +1908,10 @@
   <data name="SettingsPreferencesShowLibrarySection.Header" xml:space="preserve">
     <value>Show Library section on the sidebar</value>
   </data>
+    <data name="DefaultScheme" xml:space="preserve">
+    <value>Default</value>
+  </data>
+    <data name="SettingsApperanceColorScheme.Header" xml:space="preserve">
```
Comment (yaichenbaum) [inline Files/Strings/en-US/Resources.resw:1914]:
```suggestion
    <data name="SettingsAppearanceColorScheme.Header" xml:space="preserve">
```

File: Files/ViewModels/SettingsViewModels/AppearanceViewModel.cs [hunk 5]
```diff
@@ -139,5 +142,26 @@ public bool AreRightClickContentMenuAnimationsEnabled
...
+                    ShowRestartDialog = true;
+                }
+            }
+        }
+
+        public bool ShowRestartDialog
+        {
+            get => showRestartDialog;
+            set => SetProperty(ref showRestartDialog, value);
+        }
```
Comment (yaichenbaum) [inline Files/ViewModels/SettingsViewModels/AppearanceViewModel.cs:165]: Is it posible to swap out the themes at run time?

File: Files/Views/SettingsPages/Appearance.xaml.cs [hunk 6]
```diff
@@ -8,5 +8,7 @@ public Appearance()
         {
             InitializeComponent();
         }
+
+        public bool ShowColorSchemeSelector => App.ExternalResourcesHelper.Themes.Count > 1;
```
Comment (yaichenbaum) [inline Files/Views/SettingsPages/Appearance.xaml.cs:12]:
```suggestion
        private bool ShowColorSchemeSelector => App.ExternalResourcesHelper.Themes.Count > 1;
```

File: Files/Views/SettingsPages/Appearance.xaml [hunk 7]
```diff
@@ -43,6 +43,18 @@
                     ItemsSource="{Binding Themes}"
                     SelectedIndex="{Binding SelectedThemeIndex, Mode=TwoWay}" />
 
+                <ComboBox
+                    x:Name="ColorSchemeChooser"
+                    x:Uid="SettingsApperanceColorScheme"
```
Comment (yaichenbaum) [inline Files/Views/SettingsPages/Appearance.xaml:48]:
```suggestion
                    x:Uid="SettingsAppearanceColorScheme"
```

**Review by yair100 — CHANGES_REQUESTED** [review]

File: Files/ViewModels/SettingsViewModels/AppearanceViewModel.cs (hunk 5 above)
Comment (winston-de) [reply Files/ViewModels/SettingsViewModels/AppearanceViewModel.cs:165]: @yair100 I wasn't able to get it working, the appearance wouldn't change. I think the dictionary has to be added before the controls are rendered.

File: Files/ViewModels/SettingsViewModel.cs [hunk 8]
```diff
@@ -549,6 +549,11 @@ public bool ShowOpenInNewTabMenuItem
             set => Set(value);
         }
 
+        public string PathToThemeFile
```
Comment (yaichenbaum) [inline Files/ViewModels/SettingsViewModel.cs:552]: Can you add documentation for this setting?

**Review by yair100 — CHANGES_REQUESTED** [review]

**Review by yair100 — APPROVED** [review]
@winston-de Nice work! LGTM

**Comment by d2dyno1** [comment]
@winston-de How does a theme file look like? How do I create one myself?

**Comment by yaichenbaum** [comment]
> @winston-de How does a theme file look like? How do I create one myself?

You can find instructions here https://files-community.github.io/docs/#/articles/custom-themes.

**Pull request merged**